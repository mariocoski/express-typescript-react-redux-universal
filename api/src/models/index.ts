
import * as fs from 'fs';
import * as path from 'path';
const Sequelize = require('sequelize');
const basename  = path.basename(__filename);
const env       = process.env.NODE_ENV || 'development';
const config    = require(path.resolve(__dirname,'../config/database'))[env];

const sequelize  = new Sequelize(config.database, config.username, config.password, {...config, operatorsAliases: false, logging: false});

let db: any = {};

const files= fs.readdirSync(__dirname);
  files.filter(file => {
    return (file.indexOf('.') !== 0) && (file !== basename) && (file.slice(-3) === '.js' || file.slice(-3) === '.ts');
  }).forEach(file  => {
    const model: any = sequelize.import(path.join(__dirname, file));
    db[model.name] = model;
  });

Object.keys(db).forEach(function(modelName) {
  if ("associate" in db[modelName]) {
    db[modelName].associate(db);
  }
});

interface databaseInterface {
  sequelize: Object,
  Sequelize: Object, 
  User: Object
}
const database:databaseInterface = {sequelize, Sequelize, ...db};
module.exports = database;