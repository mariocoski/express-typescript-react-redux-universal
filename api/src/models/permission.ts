module.exports = (sequelize, DataTypes) => {
  const Permission = sequelize.define('Permission', {
    name: DataTypes.STRING,
    description: DataTypes.STRING
  }, {
    tableName: 'permissions'
  });

  Permission.associate = (models) => {
    Permission.belongsToMany(models.Role, {
      through: 'roles_permissions', foreignKey: 'permission_id'
    });
  }

  return Permission;
};