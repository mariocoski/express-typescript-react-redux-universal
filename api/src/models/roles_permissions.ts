module.exports = (sequelize: any, DataTypes: any) => {
  const RolePermission = sequelize.define('roles_permissions', {
    id : {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true
    },
    role_id: {
      type: DataTypes.INTEGER
    },
    permission_id: {
      type: DataTypes.INTEGER
    }
  },{
    tableName: 'roles_permissions',
    timestamps: false
  });
  return RolePermission;
};
