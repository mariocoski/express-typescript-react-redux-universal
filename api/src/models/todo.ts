module.exports = (sequelize: any, DataTypes: any) => {
  const Todo = sequelize.define('Todo', {
    id : {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true
    },
    user_id: {
      type: DataTypes.INTEGER,
      field: 'user_id',
    },
    title: DataTypes.STRING,
    description: DataTypes.TEXT,
    created_at: {
      type: DataTypes.DATE,
      field: 'created_at',
      defaultValue: DataTypes.NOW 
    },
    updated_at: {
      type: DataTypes.DATE,
      field: 'updated_at'
    },
    completed_at: {
      type: DataTypes.DATE,
      field: 'completed_at'
    },
    deleted_at: {
      type: DataTypes.DATE,
      field: 'deleted_at'
    }
  },{
    tableName: 'todos',
    timestamps: false
  });

   Todo.associate = (models: any) => {
    Todo.belongsTo(models.User,{
      foreignKey: {
        unique: false
      }
    });
  }

  return Todo;
};
