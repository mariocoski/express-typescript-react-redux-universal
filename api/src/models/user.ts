import * as bcrypt from 'bcrypt';
module.exports = function(sequelize, DataTypes) {
  const User = sequelize.define("User", {
    email: DataTypes.STRING,
    password: DataTypes.STRING
  });

  User.beforeCreate((user, options) => {
    return bcrypt.hash(user.password, 10)
                  .then(hash => {
                      user.password = hash;
                  })
                  .catch(err => { throw new Error(); });
  });
  
  return User;
};