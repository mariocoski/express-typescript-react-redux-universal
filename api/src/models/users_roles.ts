module.exports = (sequelize: any, DataTypes: any) => {
  const UserRole = sequelize.define('users_roles', {
    id : {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true
    },
    user_id: {
      type: DataTypes.INTEGER,
      field: 'user_id',
    },
    role_id: {
      type: DataTypes.INTEGER,
      field: 'role_id'
    }
  },{
    tableName: 'users_roles',
    timestamps: false
  });
  return UserRole;
};
