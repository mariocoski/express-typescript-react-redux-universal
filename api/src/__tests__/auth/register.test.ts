require('dotenv').config();
import {EMAIL_IS_REQUIRED, EMAIL_IS_INVALID, PASSWORD_IS_REQUIRED,
        PASSWORD_IS_TOO_SHORT} from '../../constants/errors';
import * as db from '../../models';

async function expectError(response, error){
  expect(response.statusCode).toBe(422);
  expect(response.text).toMatch(error);
}

describe('AUTH', () => {

  const request = require('supertest');
  let app;

  beforeEach((done) => {
    db['sequelize'].sync({force:true});
    app = require('../../server');
    
  });

  afterEach(() => {
    app.close();
  });

  it('should fail to create a user without input', async () => {
    expect.assertions(2);
    const response = await request(app).post('/auth/register');
    expectError(response,EMAIL_IS_REQUIRED);
  });

  it('should fail to create a user when email is invalid', async () => {
    expect.assertions(2);
    const response = await request(app).post('/auth/register')
                                       .type('form')
                                       .send( { email: 'invalid@email' });
    expectError(response,EMAIL_IS_INVALID);
  });

  it('should fail to create a user without password', async () => {
    expect.assertions(2);
    const response = await request(app).post('/auth/register')
                                       .type('form')
                                       .send( { email: 'valid@email.com' });
    expectError(response,PASSWORD_IS_REQUIRED);
  });

  it('should fail to create a user without password', async () => {
    expect.assertions(2);
    const response = await request(app).post('/auth/register')
                                       .type('form')
                                       .send( { email: 'valid@email.com', password: 'short' });
    expectError(response,PASSWORD_IS_TOO_SHORT);
  });

  it('should fail to create a user without password', async () => {
    expect.assertions(2);
    const response = await request(app).post('/auth/register')
                                       .type('form')
                                       .send( { email: 'valid@email.com', password: 'short' });
    expectError(response,PASSWORD_IS_TOO_SHORT);
  });

  // it('should create a user with valid input', async () => {
  //   expect.assertions(3);
   
  //   const response = await request(app).post('/auth/register')
  //                                      .type('form')
  //                                      .send( { email: 'valid@email.com', password: 'longenough' });
  //   const data = JSON.parse(response.text);
  //   expect(response.statusCode).toBe(201);
  //   expect(data.token).toEqual( expect.any(String));
  //   expect(data.user).toEqual( expect.any(Object));
  // });

});