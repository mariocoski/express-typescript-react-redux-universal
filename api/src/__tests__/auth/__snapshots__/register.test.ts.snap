// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REGISTER should create a user with valid input 1`] = `
Object {
  "email": "valid@email.com",
  "id": 4,
  "roles": Array [
    "user",
  ],
}
`;
